Alpha list paging implementation

smarterclayton opened the pull request:
Design in kubernetes/community#896

Support `?limit=NUMBER`, `?continue=CONTINUATIONTOKEN`, and a `continue` field
on ListMeta and pass through to etcd. Perform minor validation as an example.

```
# first out of three
$ curl http://127.0.0.1:8080/api/v1/namespaces?limit=1
{
  "kind": "NamespaceList",
  "apiVersion": "v1",
  "metadata": {
    "selfLink": "/api/v1/namespaces",
    "resourceVersion": "146",
    "next": "ZGVmYXVsdA"
  },
  "items": [
    {
      "metadata": {
        "name": "default",
        "selfLink": "/api/v1/namespaces/default",
        "uid": "f95e1390-6852-11e7-ab03-7831c1b76042",
        "resourceVersion": "4",
        "creationTimestamp": "2017-07-14T05:12:03Z"
      },
      "spec": {
        "finalizers": [
          "kubernetes"
        ]
      },
      "status": {
        "phase": "Active"
      }
    }
  ]
}
...
# last
$ curl "http://127.0.0.1:8080/api/v1/namespaces?limit=1&continue=a3ViZS1wdWJsaWM"
{
  "kind": "NamespaceList",
  "apiVersion": "v1",
  "metadata": {
    "selfLink": "/api/v1/namespaces",
    "resourceVersion": "145"
  },
  "items": [
    {
      "metadata": {
        "name": "kube-system",
        "selfLink": "/api/v1/namespaces/kube-system",
        "uid": "f95e9484-6852-11e7-ab03-7831c1b76042",
        "resourceVersion": "5",
        "creationTimestamp": "2017-07-14T05:12:03Z"
      },
      "spec": {
        "finalizers": [
          "kubernetes"
        ]
      },
      "status": {
        "phase": "Active"
      }
    }
  ]
}
```

```release-note
The Kubernetes API server now supports the ability to break large LIST calls into multiple smaller chunks.  A client can specify a limit to the number of results to return, and if more results exist a token will be returned that allows the client to continue the previous list call repeatedly until all results are retrieved.  The resulting list is identical to a list call that does not perform chunking thanks to capabilities provided by etcd3.  This allows the server to use less memory and CPU responding with very large lists.  This feature is gated as APIListChunking and is not enabled by default.  The 1.9 release will begin using this by default from all informers.
```

smarterclayton commented:
@xiang90 @wojtek-t @lavalamp @liggitt just a very rough sketch of paging in Kubernetes API

smarterclayton commented on staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go:
```diff
@@ -399,7 +401,54 @@ func (s *store) List(ctx context.Context, key, resourceVersion string, pred stor
+	var options []clientv3.OpOption
+	switch {
+	case len(fromKey) > 0 && pred.Limit > 0:
+		options = []clientv3.OpOption{clientv3.WithRev(fromRV), clientv3.WithRange(key + "\xFF"), clientv3.WithLimit(pred.Limit)}
```
@xiang90 if I want to prevent seeing updates newer than `fromRV`, what do I filter by?

smarterclayton committed:
Add a new paging utility for client side ranging

lavalamp commented:
cc @jpbetz

smarterclayton commented:
I'm still working up to testing this against our mega clusters to see if it can help bring in long tail latency.  It also potentially lets us do flyweight caches more effectively (like for GC)

smarterclayton committed:
Enable paging for all list watchers

jpbetz commented on staging/src/k8s.io/apimachinery/pkg/apis/meta/internalversion/types.go:
```diff
@@ -45,4 +46,16 @@ type ListOptions struct {
 	TimeoutSeconds *int64
+
+	From  string
+	Limit int64
```
+doc for `From` and `Limit` here for consistency with the other `ListOptions` fields?  I see they're also documented nicely in types.go.

jpbetz reviewed (commented):
Thanks for working on this @smarterclayton. I gave this a once over, focusing on conceptual stuff, not nits/code style, since this is a WIP.  Happy to review in greater detail as this matures.

smarterclayton commented on staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go:
```diff
@@ -402,16 +466,50 @@ func (s *store) List(ctx context.Context, key, resourceVersion string, pred stor
+			return err
+		}
+
+		options = append(options, clientv3.WithRange(key+"\xFF"))
```
decodeFrom performs that validation already.  The from key is a subset of the range, so there is the implicit behavior of the internal structure of the tree.  We already handle path traversal, and the mechanism here depends on this being a sorted range.  I think at some point we would offer "startFrom" (I think `From` in this PR would be renamed to `Continue` and be opaque, but `From` would be a generic tool that would have to follow a recognizable pattern).

jpbetz replied:
Thanks for clarifying. I did find `From` a bit difficult to grok at first, so agree that name like `Continue` or `NextPageToken` might be helpful.

smarterclayton commented on staging/src/k8s.io/client-go/tools/pager/pager.go:
```diff
@@ -0,0 +1,103 @@
+	PageSize int64
+	PageFn   ListPageFunc
+
+	FullListIfExpired bool
```
Pager should not implicitly allow a Continuation (valuable to allow it to be explicitly specified).  Pager hides the details of how the List is returned - clients never see the partial behavior.

If we expire during a continuation, we are very likely to expire retrying the continuation.  The fallback is for informers, which can reasonably make the decision that full list is the correct thing to do (especially controller informers).

smarterclayton replied:
Concretely, it is not intended that the consumer of this make any assumptions about what is returned by PageFn (and would definitely make it clear that PageFn may be invoked multiple times, out of order, etc).

smarterclayton commented on staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go:
```diff
@@ -328,6 +333,15 @@ type ListOptions struct {
+	From string `json:"from,omitempty" protobuf:"bytes,7,opt,name=from"`
+	// Limit is a maximum number of responses to return for a list call. If more items exist, the
+	// server will set the next field on the list metadata to a token you can pass as from.
+	Limit int64 `json:"limit,omitempty" protobuf:"varint,8,opt,name=limit"`
```
Yeah, worth discussion.  I'm not sure we support `*int64` yet in from parameters but we should.  I agree that the distinction should be preserved.

wojtek-t replied:
+1 - there definitely should be a different between those two

@smarterclayton Regarding `*int64` support - I think we do support it. E.g. "TimeoutSeconds" 10 lintes above is actually of time `*int64`.

smarterclayton commented:
I also implemented partial resource.Builder (whose Visit interface is expected to be streaming, but has never been) but it isn't part of this PR.  See https://github.com/smarterclayton/origin/commit/7a9c2967d714d0dc09b3aa1209cc595c27c861c0

That's actually extremely useful for most users - instead of multi second delays, results start streaming from get almost instantly, and very long running calls (such as client side migration logic and API conversion) can fail.  That highlighted the need for some clients to accept range expired and to be able to continue, whether by allowing direct range specification via `from` or via a special value returned by the 410 (`continueFrom` or similar).

wojtek-t commented on staging/src/k8s.io/client-go/tools/cache/listwatch.go:
```diff
@@ -87,6 +91,9 @@ func timeoutFromListOptions(options metav1.ListOptions) time.Duration {
 
 // List a set of apiserver resources
 func (lw *ListWatch) List(options metav1.ListOptions) (runtime.Object, error) {
+	if !lw.DisablePaging {
```
@smarterclayton - did you do any performance-related experiments?

E.g. if I want to list 50k of objects, what is the different with pagination for e2e time of getting the result?

wojtek-t reviewed (commented):
@smarterclayton - I took a very high level look and overall it looks great to me.

That said, I think that we should try to get the API and implementation in ASAP. I'm not sure we should enable it by default for all watchers now, but we can discuss it separately.

Please ping me if you want me to make a more careful review.

smarterclayton committed:
API for server paging

Adds a `continue` and `limit` parameter to ListOptions

smarterclayton committed:
Server side implementation of paging for etcd3

Add a feature gate in the apiserver to control whether paging can be
used. Add controls to the storage factory that allow it to be disabled
per resource. Use a JSON encoded continuation token that can be
versioned. Create a 410 error if the continuation token is expired.

Adds GetContinue() to ListMeta.

smarterclayton commented:
@sttts there is no generic implementation possible for DeepCopyObject() on List - not sure that it should be part of runtime.Object.

```
vendor/k8s.io/client-go/tools/pager/pager.go:97: cannot use list (type *internalversion.List) as type runtime.Object in return argument:
I0811 02:43:05.516] 	*internalversion.List does not implement runtime.Object (missing DeepCopyObject method)
```

sttts commented:
> edit: never mind, it just has to be hand coded.

or via tag.

porridge commented:
Does this change the semantics of `*List` `resourceVersion`? I'm just wondering why in the first example in ths PR it goes from `146` on the first page down to `145` on the last page? A typo?

smarterclayton commented:
It does not.  Might be a typo or some of the bugs I was squashing as I
prototyped.

The contract of Continue is that the resourceVersion for each chunk should
be the same as the first (described in the design proposal).  So each RV
would have to be the same and the value of the RV would be encoded in the
continue token.

On Fri, Aug 11, 2017 at 8:41 AM, Marcin Owsiany <notifications@github.com>
wrote:

> Does this change the semantics of *List resourceVersion? I'm just
> wondering why in the first example in ths PR it goes from 146 on the
> first page down to 145 on the last page? A typo?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/48921#issuecomment-321802198>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABG_p4AapMsYdJJFTGe6AHatpShoYvRIks5sXEvqgaJpZM4OX2sk>
> .
>

smarterclayton committed:
Integration test for API paging

xiang90 commented on staging/src/k8s.io/apimachinery/pkg/apis/meta/internalversion/types.go:
```diff
@@ -48,6 +47,9 @@ type ListOptions struct {
 	// Timeout for the list/watch call.
 	TimeoutSeconds *int64
+
+	Limit    int64
```
doc string?

liggitt replied:
Also, should be *int64 to distinguish 0 from unset?

smarterclayton replied:
For query params 0 and unspecified have the same behavior other than the possibility we'd reject (which I don't think we should do, since that would mean a new client running against a server with paging disabled would then fail).

liggitt replied:
I guess if we're not returning additional info like total result count there's not much reason to do a limit=0 list

smarterclayton replied:
There's client side code (that I would change by default in kube) that can't safely be enabled until we go beta.  So for the 1.8 release cycle the client side paging for informers can be on for testing, but it must be disabled before we ship 1.8 (so that if in 1.9 we realize the alpha api has to change, we don't have a big set of clients using it by default and rolling updates aren't compromised).

xiang90 commented on staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go:
```diff
@@ -82,23 +86,24 @@ type objState struct {
 // where Get operations don't require quorum read.
-func NewWithNoQuorumRead(c *clientv3.Client, codec runtime.Codec, prefix string, transformer value.Transformer) storage.Interface {
-	return newStore(c, false, codec, prefix, transformer)
+func NewWithNoQuorumRead(c *clientv3.Client, codec runtime.Codec, prefix string, transformer value.Transformer, pagingEnabled bool) storage.Interface {
```
nit: bool arg on public func is not preferred. it is not readable.

smarterclayton replied:
Yeah, don't have any easy way to refactor.  We should refactor in the future.

xiang90 commented on staging/src/k8s.io/apiserver/pkg/storage/storagebackend/factory/etcd3.go:
```diff
@@ -61,7 +61,7 @@ func newETCD3Storage(c storagebackend.Config) (storage.Interface, DestroyFunc, e
+		return etcd3.New(client, c.Codec, c.Prefix, transformer, c.Paging), destroyFunc, nil
 	}
-	return etcd3.NewWithNoQuorumRead(client, c.Codec, c.Prefix, transformer), destroyFunc, nil
+	return etcd3.NewWithNoQuorumRead(client, c.Codec, c.Prefix, transformer, c.Paging), destroyFunc, nil
```
probably we should simply pass config into etcd3.New instead of separated field :P

smarterclayton replied:
Yeah, this code needs a good hard refactor.  But I don't want to do it here, and I wouldn't stop just at the factory (the three levels above that are not well composed right now).

xiang90 replied:
sure. agreed.

xiang90 commented on staging/src/k8s.io/apiserver/pkg/storage/interfaces.go:
```diff
@@ -37,7 +37,7 @@ type Versioner interface {
 	// cannot be updated correctly. May return nil if the requested object does not need metadata
 	// from database.
-	UpdateList(obj runtime.Object, resourceVersion uint64) error
+	UpdateList(obj runtime.Object, resourceVersion uint64, next string) error
```
we need to either:

1. document next is optional
2. or make next an option arg by doing (obj runtime.Object, resourceVersion uint64, opts listopts...)

xiang90 commented:
looks good to me. thanks!

wojtek-t commented on staging/src/k8s.io/apiserver/pkg/storage/selection_predicate.go:
```diff
@@ -76,6 +76,8 @@ type SelectionPredicate struct {
 	IncludeUninitialized bool
 	GetAttrs             AttrFunc
 	IndexFields          []string
+	Limit                int64
```
Seems unused now.

smarterclayton replied:
Should be used in the `List(` method

wojtek-t reviewed (commented):
I skimmed through the code once again and it looks ok to me.

Feel free to not wait for me (though I will try to review it over the next few days).

smarterclayton commented:
/test pull-kubernetes-e2e-kops-aws

smarterclayton commented:
/test pull-kubernetes-e2e-kops-aws

wojtek-t commented on staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/generated.proto:
```diff
@@ -320,6 +320,14 @@ message ListMeta {
+  // continue may be set if the user set a limit on the number of items returned, and indicates that
+  // the server has more data available. The value is opaque and may be used to issue another request
+  // to the endpoint that served this list to retrieve the next set of available objects. Continuing a
+  // list may not be possible if the server configuration has changed or more than a few minutes have
```
Is this few minutes somehow defined? I think the actual time that needs to elapse may be different.
So I wouldn't be that specific saying about minutes.

WDYT?

smarterclayton replied:
I think the problem is, if we do not specify some reasonable expectation here API consumers will likely not have any expectation.  We should be clear that this may be changed per consumer, but setting a rough magnitude is important.

Continue depends on historical MVCC values, which depends on compaction interval.  It's arbitrarily set to 5 minutes today.  Rate of change in the cluster creates dead values which increases resident size on disk and in memory, and compaction clears those and also makes historical reads not possible.  In the future we may tune this up.

wojtek-t replied:
But isn't it that it may significantly vary?
If you start you paginated LIST right before the compation, you will have the data available for just very small amount of time, right? Or am i missing something?

If so, this on average will be in the order of magnitude of minutes, but can be either 1s or <compaction interval>. So it's not hard guarantee.

Or am I completely missing something?

liggitt replied:
Is compaction not rolling? I thought it compacted up to a fixed time in the past, not all events up to the present

liggitt replied:
Yeah, compaction keeps x minutes of history

wojtek-t replied:
Ahh ok. Sorry. I forgot about it.
I think you're right that we are compacting up to "previous time when we did compaction". So indeed, with compaction happening every 5 minutes, the history time when you LIST will still be valid should be something between 5 and 10 minutes.

wojtek-t commented on staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go:
```diff
@@ -333,6 +341,33 @@ type ListOptions struct {
+	// is server defined, clients may only use the continue value from a previous query result with
+	// identical query parameters (except for the value of continue) and the server may reject a continue
+	// value it does not recognize. If the specified continue value is no longer valid whether due to
+	// expiration (generally five to fifteen minutes) or a configuration change on the server the server
```
Again - I would avoid mentioning this "five to fifteen minutes" here. Even though it may be true for etcd (btw - what is triggering this - churn on the cluster?), if at some point we will have a different backing storage, it may be different.
So I would avoid putting this information in the API definition.

wojtek-t commented on staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go:
```diff
@@ -386,7 +391,62 @@ func (s *store) GetToList(ctx context.Context, key string, resourceVersion strin
+// continueToken is a simple structured object for encoding the state of a continue token.
+// TODO: if we change the version of the encoded from, we can't start encoding the new version
+// until all other servers are upgraded (i.e. we need to support rolling schema)
+// TODO: be very careful about changing this
```
Should this be a TODO? It sounds like a real comment.

smarterclayton replied:
Yeah, will update.

wojtek-t commented on staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go:
```diff
@@ -386,7 +391,62 @@ func (s *store) GetToList(ctx context.Context, key string, resourceVersion strin
+		if len(c.StartKey) == 0 {
+			return "", 0, fmt.Errorf("continue key is not valid: encoded start key empty (version v1)")
+		}
+		cleaned := path.Clean(c.StartKey)
```
I'm wondering about this one. Are the paths that are "Clean-equivalent" really equivalent in etcd?

smarterclayton replied:
For the purposes of this code, it's mostly ensuring no path traversal attacks are possible.  If the start key is `foo/bar/baz` and this is a cluster scoped resource, it doesn't matter.  If an attacker passes `/`, we still append to the end of the current prefix.

wojtek-t replied:
Got it - can you please add a comment about it. Your goal definitely wasn't clear to me before you explained it.

serathius replied:
This cannot happen as etcd doesn't have idea of path. Key is just a blob, it's Kubernetes that puts paths in key.

enj replied:
etcd may not have paths, but by using slashes and prefixing, we have paths and traversal attacks #113756

serathius replied:
This applies to apiserver HTTP server that will interpret `/safePath/../unsafePath` as `/unsafePath`. However etcd doesn't have concept of path and will interpret `/safePath/../unsafePath` as `/safePath/../unsafePath`.

So no, etcd is not vulnerable to traversal attacks.

wojtek-t commented on staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go:
```diff
@@ -386,7 +391,62 @@ func (s *store) GetToList(ctx context.Context, key string, resourceVersion strin
+		if cleaned != c.StartKey || cleaned == "." || cleaned == "/" {
+			return "", 0, fmt.Errorf("continue key is not valid: %s", cleaned)
+		}
+		if len(cleaned) == 0 {
```
Can't we merge this if with the one above?

smarterclayton replied:
Could, but I wanted a specific message for the user

wojtek-t reviewed (commented):
I finally got to review. I just had few minor comments.

@smarterclayton - please fix the "verify" script and given that @xiang90 already reviewed it, I'm happy to LGTM it after that.

smarterclayton commented:
Updated with latest comments.  

Note that this PR will enable clients sending chunking info, and we will get some soak time in master with all test suites on, then we will disable client chunking prior to release (probably next week).  Because client side paging is defined in every ListWatch in the codebase, it's impractical to change the public signature of all generated code, and it cannot rely on features because it is in client-go.  We will not ship 1.8 with clients sending the chunking parameters by default.

smarterclayton commented:
Other review comments?

smarterclayton committed:
generated: api changes

wojtek-t commented:
My only comment is about the comment saying that it's in order of fifteen minutes or so, I would like to get rid of it. But I'm not going to block on it.

@smarterclayton please fix the verify script and I'm happy to lgtm it then.

smarterclayton commented:
I really really want to have a rough bound on time, we can always adjust it up or make it differently.  But it's going to be 5 ootb and I don't want people assuming it is forever.

wojtek-t commented:
> and I don't want people assuming it is forever.

That one I completely agree with. I just didn't want to mention "five to fifteen minutes" in the comment. 

But well, let's not waste too much time on this discussion.

/lgtm

smarterclayton commented:
/retest

smarterclayton commented:
/retest

wojtek-t commented:
@smarterclayton - this requires a rebase (not sure why bot didn't mention it).

smarterclayton commented:
Trivial rebase

smarterclayton commented:
/retest
